File: backend/pipelines/step2_preprocessing/validity.py
from typing import List, Dict
import numpy as np

from .face_filter import detect_faces
from .quality_checks import is_frame_too_dark, is_frame_blurry


def build_validity_mask_with_stats(frames: List[np.ndarray]) -> Dict:
    face_counts = detect_faces(frames)

    valid_mask = []

    stats = {
        "total_frames": len(frames),
        "valid_frames": 0,
        "no_face_frames": 0,
        "multi_face_frames": 0,
        "dark_frames": 0,
        "blurry_frames": 0,
    }

    for frame, n_faces in zip(frames, face_counts):
        valid = True  # RESET every frame

        # Face rules
        if n_faces == 0:
            stats["no_face_frames"] += 1
            valid = False
        elif n_faces > 1:
            stats["multi_face_frames"] += 1
            valid = False

        # Lighting rule
        if is_frame_too_dark(frame):
            stats["dark_frames"] += 1
            valid = False

        # Blur rule (secondary)
        if is_frame_blurry(frame) and n_faces != 1:
            stats["blurry_frames"] += 1
            valid = False

        if valid:
            stats["valid_frames"] += 1

        valid_mask.append(valid)
        

    return {
        "valid_mask": valid_mask,
        "stats": stats
    }


def evaluate_video_quality(
    valid_mask: List[bool],
    stats: Dict,
    min_valid_frames: int = 100,
    min_valid_ratio: float = 0.7,
    max_invalid_gap_sec: float = 3.0,
    fps: int = 25
) -> Dict:

    # Guard: no frames at all
    if stats["total_frames"] == 0:
        return {"usable": False, "reason": "NO_FRAMES_DECODED"}

    # Rule 1: too few valid frames (YOUR REQUEST)
    if stats["valid_frames"] < min_valid_frames:
        return {
            "usable": False,
            "reason": "TOO_FEW_VALID_FRAMES",
        }

    # Rule 2: valid ratio
    valid_ratio = stats["valid_frames"] / stats["total_frames"]
    if valid_ratio < min_valid_ratio:
        return {
            "usable": False,
            "reason": "LOW_VALID_RATIO",
        }

    # Rule 3: long continuous invalid gap
    max_gap = 0
    current_gap = 0
    for v in valid_mask:
        if not v:
            current_gap += 1
            max_gap = max(max_gap, current_gap)
        else:
            current_gap = 0

    if max_gap / fps > max_invalid_gap_sec:
        return {
            "usable": False,
            "reason": "LONG_INVALID_GAP",
        }

    return {"usable": True, "reason": "OK"}

File: frontend/lib/validity.ts
import { isFrameTooDark, isFrameBlurry } from "./qualityChecks";
import {
  sendFaceFrame,
  getLatestFaceCount,
} from "./faceDetection";

export type ValidityStats = {
  total_frames: number;
  valid_frames: number;
  no_face_frames: number;
  multi_face_frames: number;
  dark_frames: number;
  blurry_frames: number;
};

export function createInitialStats(): ValidityStats {
  return {
    total_frames: 0,
    valid_frames: 0,
    no_face_frames: 0,
    multi_face_frames: 0,
    dark_frames: 0,
    blurry_frames: 0,
  };
}

/**
 * Non-blocking frame validity processing
 */
export function processFrameValidity(
  video: HTMLVideoElement,
  canvas: HTMLCanvasElement,
  stats: ValidityStats,
  validMask: boolean[]
) {
  if (
    video.videoWidth === 0 ||
    video.videoHeight === 0
  ) {
    return;
  }

  const ctx = canvas.getContext("2d");
  if (!ctx) return;

  canvas.width = video.videoWidth;
  canvas.height = video.videoHeight;

  ctx.drawImage(video, 0, 0);

  const imageData = ctx.getImageData(
    0,
    0,
    canvas.width,
    canvas.height
  );

  stats.total_frames += 1;

  let valid = true;

  // 🔹 Face detection (event-driven)
  sendFaceFrame(video);
  const faceCount = getLatestFaceCount();

  if (faceCount === 0) {
    stats.no_face_frames += 1;
    valid = false;
  } else if (faceCount > 1) {
    stats.multi_face_frames += 1;
    valid = false;
  }

  // 🔹 Lighting
  if (isFrameTooDark(imageData)) {
    stats.dark_frames += 1;
    valid = false;
  }

  // 🔹 Blur (secondary rule, same as Python)
  if (
    isFrameBlurry(
      imageData,
      canvas.width,
      canvas.height
    ) &&
    faceCount !== 1
  ) {
    stats.blurry_frames += 1;
    valid = false;
  }

  if (valid) {
    stats.valid_frames += 1;
  }

  validMask.push(valid);
}

/**
 * Equivalent of evaluate_video_quality() from Python
 */
export function evaluateVideoQuality(
  validMask: boolean[],
  stats: ValidityStats,
  minValidFrames: number = 50,
  minValidRatio: number = 0.7,
  maxInvalidGapSec: number = 3.0,
  fps: number = 15
): { usable: boolean; reason: string } {
  if (stats.total_frames === 0) {
    return { usable: false, reason: "NO_FRAMES_DECODED" };
  }

  if (stats.valid_frames < minValidFrames) {
    return {
      usable: false,
      reason: "TOO_FEW_VALID_FRAMES",
    };
  }

  const validRatio =
    stats.valid_frames / stats.total_frames;

  if (validRatio < minValidRatio) {
    return {
      usable: false,
      reason: "LOW_VALID_RATIO",
    };
  }

  let maxGap = 0;
  let currentGap = 0;

  for (const v of validMask) {
    if (!v) {
      currentGap += 1;
      maxGap = Math.max(maxGap, currentGap);
    } else {
      currentGap = 0;
    }
  }

  if (maxGap / fps > maxInvalidGapSec) {
    return {
      usable: false,
      reason: "LONG_INVALID_GAP",
    };
  }

  return { usable: true, reason: "OK" };
}
